src/Pretzel/Commands/CommandCollection.cs

```diff
@@ -12,56 +12,72 @@ namespace Pretzel.Commands
     [Shared]
     public sealed class CommandCollection
     {
+        ExportFactory<Logic.Commands.ICommand, CommandInfoAttribute>[] commands;
         [ImportMany]
-        public ExportFactory<ICommand, CommandInfoAttribute>[] Commands { get; set; }
-        [ImportMany]
-        public IEnumerable<Lazy<IHaveCommandLineArgs>> CommandLineExtensions { get; set; }
-        [Import]
-        public Lazy<CommandParameters> Parameters { get; set; }
+        public ExportFactory<Logic.Commands.ICommand, CommandInfoAttribute>[] Commands
+        {
+            get => commands ?? Array.Empty<ExportFactory<Logic.Commands.ICommand, CommandInfoAttribute>>();
+            set => commands = value;
+        }
 
-        private Dictionary<string, ICommand> commandMap;
+        ICommandArguments[] commandArguments;
+        [ImportMany]
+        public ICommandArguments[] CommandArguments
+        {
+            get => commandArguments ?? Array.Empty<ICommandArguments>();
+            set => commandArguments = value;
+        }
 
-        public ICommand this[string name]
+        ExportFactory<ICommandArgumentsExtension, CommandArgumentsExtensionAttribute>[] argumentExtensions;
+        [ImportMany]
+        public ExportFactory<ICommandArgumentsExtension, CommandArgumentsExtensionAttribute>[] ArgumentExtensions
         {
-            get
-            {
-                ICommand command;
-                commandMap.TryGetValue(name.ToLower(System.Globalization.CultureInfo.InvariantCulture), out command);
-                return command;
-            }
+            get => argumentExtensions ?? Array.Empty<ExportFactory<ICommandArgumentsExtension, CommandArgumentsExtensionAttribute>>();
+            set => argumentExtensions = value;
         }
 
+        [Export]
+        public RootCommand RootCommand { get; set; }
+
+        [Import]
+        public IConfiguration Configuration { get; set; }
+
         [OnImportsSatisfied]
-        public void OnImportsSatisfied()
+        internal void OnImportsSatisfied()
         {
-            commandMap = new Dictionary<string, ICommand>(Commands.Length);
+            RootCommand = new RootCommand();
 
             foreach (var command in Commands)
             {
-                if (!commandMap.ContainsKey(command.Metadata.CommandName))
+                var subCommand = new Command(command.Metadata.CommandName, command.Metadata.CommandDescription);
```

Comment:
I think we can simplify all this by exporting directly `Command`s: each command will inherit from `Command` and thus 
handle their options.
This way we can just `ImportMany` commands and add them to the `RootCommand`.

Regarding the arguments extensibility, we can use the interface `ICommandArgumentsExtension` with a property or method indicating the type of the command: `public Type For() => typeof(BakeCommand);`
I wanted to use a generic parameter but I haven't found a way to export it correctly.

Reply:
I did not want to do that for a reason: If we import the `commands` directly we import all dependencies with it. If a plugin provides a custom command, but is broken in some kind of way (for example it's incompatible after a new release of pretzel) it will break the whole application, even the custom command is not called by the end user.

The way it is implemented right now does instanciate as less as possible via MEF before we know what command the end user picked.

Reply:
> Regarding the arguments extensibility, we can use the interface ICommandArgumentsExtension with a property or method indicating the type of the command: public Type For() => typeof(BakeCommand);
I wanted to use a generic parameter but I haven't found a way to export it correctly.

That is possible, but then we need to move the commands from `Petzel` to `Pretzel.Logic` and make them public. That way we can drop the `IBakeCommandArguments`. That's a reasonable change in my opinion cause we don't throw around all those argument interfaces.

Reply:
> I did not want to do that for a reason: If we import the `commands` directly we import all dependencies with it. If a plugin provides a custom command, but is broken in some kind of way (for example it's incompatible after a new release of pretzel) it will break the whole application, even the custom command is not called by the end user.

I understand your view but in the case of this small app I don't think it's worth so much complexity: if a plugin is failing the use will necessarily have to replace or update it since it was needed by the user.

> The way it is implemented right now does instanciate as less as possible via MEF before we know what command the end user picked.

I think we can delay and handle the exception through lazy imports.
 
> If we use the change you suggested, we will need to instanciate the commands or do some kind of reflection trickery, instead of using the path `MEF` provides.

Sorry but I don't understand why we will need to instanciate the commands, they will be available directly through the import.
You can see an example in my testing repo: the [custom command](https://github.com/laedit/CommandLineExtensibilityWithMEF2/blob/master/CommandLineExtensibility/BuiltinCommand2.cs) is [exported as a command](https://github.com/laedit/CommandLineExtensibilityWithMEF2/blob/master/CommandLineExtensibility/Program.cs#L104) (through convention but the same can be achieved through attributes) and is then [imported as command](https://github.com/laedit/CommandLineExtensibilityWithMEF2/blob/master/CommandLineExtensibility/Starter.cs#L60).
Or am I missing something?


> That is possible, but then we need to move the commands from `Petzel` to `Pretzel.Logic` and make them public. That way we can drop the `IBakeCommandArguments`. That's a reasonable change in my opinion cause we don't throw around all those argument interfaces and can just export the arguments directly.

Good to me.

Reply:
> > I did not want to do that for a reason: If we import the `commands` directly we import all dependencies with it. If a plugin provides a custom command, but is broken in some kind of way (for example it's incompatible after a new release of pretzel) it will break the whole application, even the custom command is not called by the end user.
> 
> I understand your view but in the case of this small app I don't think it's worth so much complexity: if a plugin is failing the use will necessarily have to replace or update it since it was needed by the user.

I think this from an end users perspective (I know its hypothetical for now but realistic in my opinion):

A users starts using pretzel. He want's to start his blog cause he is frustrated using his current blog software for example medium.

We have a plugin `Pretzel.Plugins.MediumImporter`.
He pulls in the dependency, imports the blog:

`pretzel import medium --foo --bar`
runs
`pretzel taste`

is happy blogging all a long, never needs the import plugin again.

We release a version, plugin stops working for whatever reason.
Suddenly `pretzel taste` will break. Enduser is puzzeled why it stopped working.

Another reason is: Almost every dependency in the whole app get's instanciated, that's wasteful.

But maybe I'm to parnoid this time 😅


> 
> > The way it is implemented right now does instanciate as less as possible via MEF before we know what command the end user picked.
> 
> I think we can delay and handle the exception through lazy imports.

That's a possibility. Still persists the problem of instanciating all the dependencies in the app. as mentioned above.

> 
> > If we use the change you suggested, we will need to instanciate the commands or do some kind of reflection trickery, instead of using the path `MEF` provides.
> 
> Sorry but I don't understand why we will need to instanciate the commands, they will be available directly through the import.
> You can see an example in my testing repo: the [custom command](https://github.com/laedit/CommandLineExtensibilityWithMEF2/blob/master/CommandLineExtensibility/BuiltinCommand2.cs) is [exported as a command](https://github.com/laedit/CommandLineExtensibilityWithMEF2/blob/master/CommandLineExtensibility/Program.cs#L104) (through convention but the same can be achieved through attributes) and is then [imported as command](https://github.com/laedit/CommandLineExtensibilityWithMEF2/blob/master/CommandLineExtensibility/Starter.cs#L60).
> Or am I missing something?

If you import something it will get instanciated. If we import lazy we have still no way to get to the arguments from the command without instanciating it. 🤷‍♂️

> 
> > That is possible, but then we need to move the commands from `Petzel` to `Pretzel.Logic` and make them public. That way we can drop the `IBakeCommandArguments`. That's a reasonable change in my opinion cause we don't throw around all those argument interfaces and can just export the arguments directly.
> 
> Good to me.

Good let's do this 👍

Reply:
I see your point and I admit I have never thought too much about plugin broked or instantiate all the things since I have inherited the project.

So, ok to protect the end-user on this, but my only regret in this is the complexity to relate a command and his arguments.
Is it possible to "merge" the command and his arguments?
Maybe with an interface with:
- a method returning his options
- an execute method
And keeping the CommandInfo attribute with the name and description.

Reply:
I am open on this, and I tought a lot about it. I think having a clear boundary between arguments and commands is a good architecture. 
I can refactor it in this shape, but let's see how it looks like after the move to pretzel.logic. 

I still think it's a good idea to keep it appart, cause it was really hard to get it working (and the tests are tons of times easier to reason about) but you are the maintainer of the project, so you have the last word

Reply:
Another thing is extensibility, if we keep the arguments separate, it's easier to extend. But it's more or less a design choice

Reply:
Another thing is extensibility, if we keep the arguments separate, it's easier to extend. But it's more or less a design choice

Reply:
I moved the commands to logic and removed the arguments interfaces.

Reply:
Thanks, it has really helped me to get out of the fog I was regarding your proposal.
So, ok for command / arguments separation.

There is still a thing that bothers me, sorry for noticing it only now, but the command arguments extensibility have shifted his paradigm, before we were [extending a command](https://github.com/Code52/pretzel/blob/master/src/Pretzel.Logic/Extensibility/Extensions/AzureHostSupport.cs#L33) and now we are [extending a command arguments](https://github.com/biohazard999/pretzel/blob/topic/System.CommandLine.Experimental-346/src/Pretzel.Logic/Extensibility/Extensions/AzureHostSupport.cs#L14).
To me it simple more intuitive to extend a command than a command's arguments.
Is the change from a technical reason or is it from your point of view?

Reply:
Basically it's the same story as above.

You can extend an command without the need of a specifying an additional argument.
I don't know which command is extended without instanciating the command with its extentions.

Therefore: As an plugin author: If you like to provide additional command line arguments you extend the arguments. If you extend the command/engine/whatever you import the additional arguments and extend pretzel.

But basically nothing hinders the plugin author to implement `ICommandArgumentsExtension` on the extentsion it self:

```cs
[Export]
    [Shared]
    [CommandArgumentsExtension(CommandArgumentTypes = new[] { typeof(IRecipeCommandArguments) })]
    [Export(typeof(IAdditionalIngredient))]
    public class AzureHostSupport : IAdditionalIngredient, ICommandArgumentsExtension
    {
        public IList<Option> Options { get; } = new[]
        {
            new Option(new [] { "--azure", "-azure" }, "Enables deploy to azure support")
            {
                Argument = new Argument<bool>()
            }
        };

        public void BindingCompleted()
        {
            //Not used
        }
        public bool Azure { get; set; }

        private readonly IFileSystem fileSystem;
        private readonly IAssembly assembly;

        [ImportingConstructor]
        public AzureHostSupport(IFileSystem fileSystem, IAssembly assembly)
        {
            this.fileSystem = fileSystem;
            this.assembly = assembly;
        }

        public void MixIn(string directory)
        {
            if (!Azure) return;
            // Move everything under the _source folder
            var sourceFolder = Path.Combine(directory, "_source");
            if (!fileSystem.Directory.Exists(sourceFolder))
            {
                fileSystem.Directory.CreateDirectory(sourceFolder);
            }

            foreach (var file in fileSystem.Directory.GetFiles(directory))
            {
                var trimStart = file.Replace(directory, string.Empty).TrimStart(Path.DirectorySeparatorChar);
                fileSystem.File.Move(file, Path.Combine(sourceFolder, trimStart));
            }

            foreach (var directoryToMove in fileSystem.Directory.GetDirectories(directory).Where(n => new DirectoryInfo(n).Name != "_source"))
            {
                var trimStart = directoryToMove.Replace(directory, string.Empty).TrimStart(Path.DirectorySeparatorChar);
                fileSystem.Directory.Move(directoryToMove, Path.Combine(sourceFolder, trimStart));
            }

            fileSystem.File.WriteAllText(Path.Combine(directory, @"Shim.cs"), Properties.RazorAzure.Shim);
            fileSystem.File.WriteAllText(Path.Combine(directory, @"Shim.csproj"), Properties.RazorAzure.ShimProject);
            fileSystem.File.WriteAllText(Path.Combine(directory, @"Shim.sln"), Properties.RazorAzure.ShimSolution);

            var currentPath = assembly.GetEntryAssemblyLocation();
            var destination = Path.Combine(directory, "Pretzel.exe");
            if (!fileSystem.File.Exists(destination))
            {
                fileSystem.File.Copy(currentPath, destination);
            }

            Tracing.Info("Shim project added to allow deployment to azure websites");
        }
    }
```

How a command is extensible is decided by the command it self. 

Until the point where the command is created only arguments exist. 
They provide information about possible values, descriptions, help text, argument validation etc.

Hope that makes sense.

Reply:
Of course we could change the `CommandArgumentsExtensionAttribute` to accept the `typeof(BakeCommand)` instead of `typeof(BakeCommandArguments)` cause we have all the metadata on the command to find the correct argument to extend.

Reply:
> Of course we could change the `CommandArgumentsExtensionAttribute` to accept the `typeof(BakeCommand)` instead of `typeof(BakeCommandArguments)` cause we have all the metadata on the command to find the correct argument to extend.

If that is possible without questioning the rest of your work please do it, otherwise leave as it is.
Again thanks a lot for your fabulous work! 😄

Reply:
Okay, now it's possible, but i need to add the command type to the `CommandInfoAttribute` (I don't know why I didn't come up with this idea earlier). It's ugly and redundant as hell (I think that was the reason) but I hope we fix this with #353

Reply:
I was thinking the same, but yes I think we can automatically export the command with a metadata including his type.
If that's not possible and if that is ok with you, we will rollback only for this attribute property and uses only the CommandArguments for extensibility.